File: server/scripts/dynamo-contract.ts
// Runs the store contract against DynamoDB Local. Used by `make it`, not `make check`.
// FAT_HORSES_DYNAMODB_ENDPOINT defaults to http://localhost:8000.

import { DeleteTableCommand, DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { runContract, CONTRACT_SCENARIOS } from "../src/store/contract";
import { createTable, DynamoStore } from "../src/store/dynamo";

const config = {
  endpoint: process.env.FAT_HORSES_DYNAMODB_ENDPOINT ?? "http://localhost:8000",
  region: "ap-southeast-2",
  credentials: { accessKeyId: "local", secretAccessKey: "local" },
};
const client = new DynamoDBClient(config);
const tables: string[] = [];

await runContract(async () => {
  const name = `contract-${process.pid}-${tables.length}`;
  await createTable(config, name);
  tables.push(name);
  return new DynamoStore(name, client);
});
for (const t of tables) await client.send(new DeleteTableCommand({ TableName: t })).catch(() => undefined);
console.log(`dynamo contract: OK (${CONTRACT_SCENARIOS.length} scenarios)`);

File: crates/store/examples/dynamo_contract.rs
//! Runs the store contract against DynamoDB Local. Used by `make it`, not `make check`.
//!
//! Needs `FAT_HORSES_DYNAMODB_ENDPOINT` (default `http://localhost:8000`).

use std::sync::atomic::{AtomicUsize, Ordering};

use aws_sdk_dynamodb::Client;
use aws_sdk_dynamodb::config::{Credentials, Region};
use fat_horses_store::contract;
use fat_horses_store::dynamo::{DynamoStore, create_table};

/// One table per contract scenario, so each starts empty.
const TABLES: usize = 20;

#[tokio::main]
async fn main() {
    let endpoint = std::env::var("FAT_HORSES_DYNAMODB_ENDPOINT")
        .unwrap_or_else(|_| "http://localhost:8000".into());
    let config = aws_sdk_dynamodb::Config::builder()
        .behavior_version_latest()
        .endpoint_url(&endpoint)
        .region(Region::new("ap-southeast-2"))
        .credentials_provider(Credentials::new("local", "local", None, None, "static"))
        .build();
    let client = Client::from_conf(config);

    let run = std::process::id();
    let names: Vec<String> = (0..TABLES).map(|i| format!("contract-{run}-{i}")).collect();
    for name in &names {
        create_table(&client, name).await.expect("create table");
    }
    let next = AtomicUsize::new(0);
    contract::run_all(|| {
        let i = next.fetch_add(1, Ordering::SeqCst);
        DynamoStore::new(client.clone(), names[i].clone())
    })
    .await;
    for name in &names {
        let _ = client.delete_table().table_name(name).send().await;
    }
    println!(
        "dynamo contract: OK ({} scenarios)",
        next.load(Ordering::SeqCst)
    );
}
